document notification2 token refresh, unsubscribe MTM-42417 (#126)
SoftwareAG/c8y-docs: content/reference/notifications-bundle/overview.md

For Java developers, the API and the protocol have been wrapped up as an open Java API and a sample WebSocket client application.

There is a sample microservice available in the [cumulocity-examples repository](https://github.com/SoftwareAG/cumulocity-examples/tree/develop/hello-world-notification-microservice), so Java developers do not need to code to the following protocol specification directly.

> ### Token expiry

Reviewer: here and more occurrences: "expiry" -> "expiration"

Author: British english / Archaic. I'll also spell check as I'm editing this as raw text.

Author: done

### Token expiration

When creating a token, an expiration time must be given in minutes of validity from when the token was created.
This security feature limits the potential damage due to leaking of a token.
It requires tokens to be re-created or refreshed periodically.
This can be done by calling token create with the same parameters as originally. 
If the parameters used are not available they can be extracted from the token.
As the token string is a JWT (JSON Web Token), it can be decoded to extract the original information used to create the token by splitting it into 3 parts (on ".") and doing a base64 decode on the first substring.
This way, information like the subscription name can be extracted and the create token REST point can be called again, all on the client side.
The {{< product-c8y-iot >}} microservice Java SDK [TokenApi](https://github.com/SoftwareAG/cumulocity-clients-java/blob/develop/java-client/src/main/java/com/cumulocity/sdk/client/messaging/notifications/TokenApi.java) class contains a public refresh method which is implemented purely on the client side.

> ### Shared subscription

Reviewer 2: I know we have an implementation for `isShared` but I don't recall if we have completed testing for it yet?  If not then we shouldn't push the doc for it until the testing is done and we're ready for the change to (in principle) go to customers.

Author: Yes, it's doc first for once. We have a JIRA for the unsubscribe and a PR for enabling Shared and Volatile which we had there in disabled form but they require more testing (outside of the dev testing I've done so far).

Reviewer 3: I would prefer we just raise a separate PR for the token refresh for now.

Author: it can wait as it's just explaining. Maybe with unsubscribe if volatile/shared is internal only for now.

Reviewer 2: Definitely happy to see doc being written early!  I'm still going to be annoying and say that this should be split into several PRs though.

Author: I've split out shared and volatile to separate review now.

The application can consist of more than one instance running in parallel in the shared use case.

Notifications will be delivered in order with respect to the notification generating device.
The notifications will be delivered to the same application instance, except when a new instance or a failure changes the application topology.
Then it is necessary to re-distribute the devices to currently running instances.
In order to aid this assignment, the subscriber instance can specify a "consumer name" when connecting to the web socket endpoint.
The same token, or one that was generated in a similar way from subscription name and subscriber, is used as the token query string argument.
However, a different consumer name is passed as the "consumer" query string argument.
For example, instance 1 of the subscriber microservice could pass in `notification2/consumer?token=xyz&consumer=instance1` while instance 2 could use `notification2/consumer?token=XYZ&consumer=instance2`.
Currently, determining the instance ID of a microservice replica is not supported by the {{< product-c8y-iot >}} microservice API. An external application with named instances can be used instead.

> ### Volatile subscriptions

Reviewer 2: As for `isShared` - is this fully tested yet?

When subscribing, it is possible to pass in an optional boolean "isVolatile" query parameter set to true.

> Note that there is no need to mark a subscription as shared - only the token is marked shared, but here it's both (token has isVolatile and isShared).

Reviewer 2: Not sure I follow this... what is the "here" in "here it's both" referring to?

This changes the subscription to not persist notifications for the named subscription.
They are effectively only buffered in memory and can be discarded if they are not consumed quickly enough or on node failure.
Note that there can be both volatile and ordinary (non-volatile or persistent) subscriptions on a managed object with the same subscription name.

> These count as separate subscriptions and can be consumed by a subscriber using a token with the corresponding `isVolatile` equals true or false.   

Reviewer 2: "equals" --> "equal to"

> ### Unsubscribing a subscriber

Reviewer 2: Again, I don't think we can include this documentation until the implementation and testing are also done.

Once a subscription is made, notifications will be kept until consumed by all subscribers who have previously connected to the subscription.
For non-volatile subscriptions, this can result in notifications building up in storage if never consumed by the application.
These will be deleted if a tenant is deleted but otherwise can take up considerable space in permanent storage for high frequency notification sources.
It is therefore advisable to unsubscribe a subscriber that will never run again.

> A separate REST endpoint is avaiable for this: <kbd>/notification2/unsubscibe</kbd>.

Reviewer 2: "avaiable" --> "available"

For Java developers, the API and the protocol have been wrapped up as an open Java API and a sample WebSocket client application.

There is a sample microservice available in the [cumulocity-examples repository](https://github.com/SoftwareAG/cumulocity-examples/tree/develop/hello-world-notification-microservice), so Java developers do not need to code to the following protocol specification directly.

### Token expiration

When creating a token, an expiration time must be given in minutes of validity from when the token was created.
This security feature limits the potential damage due to leaking of a token.
It requires tokens to be re-created or refreshed periodically.

> This can be done by calling token create with the same parameters as originally. 

Reviewer: is "token create" a command? if so, we could use backticks to make it monospace font.

Author: Thanks for all these comments Richard. Unfortunately, it's still a bit of a work in progress but it allows us to have the text to fold in when we next release.

If the parameters used are not available they can be extracted from the token.
As the token string is a JWT (JSON Web Token), it can be decoded to extract the original information used to create the token by splitting it into 3 parts (on ".") and doing a base64 decode on the first substring.
This way, information like the subscription name can be extracted and the create token REST point can be called again, all on the client side.
The {{< product-c8y-iot >}} microservice Java SDK [TokenApi](https://github.com/SoftwareAG/cumulocity-clients-java/blob/develop/java-client/src/main/java/com/cumulocity/sdk/client/messaging/notifications/TokenApi.java) class contains a public refresh method which is implemented purely on the client side.

### Shared subscription

> There can be multiple subscription on a managed object, each receiving filtered notifications as specified by their individual subscriptions.

Reviewer: ```suggestion
There can be multiple subscriptions on a managed object, each receiving filtered notifications as specified by their individual subscriptions.
```
the sentence reads like there are subscriptions of subscriptions. is that correct?

> In order to scale, shared subscriptions are required so that notifications are dispatched to one of a number of possible consumers that are part of the same logical subscriber or parallelised application.

Reviewer: ```suggestion
In order to scale, shared subscriptions are required so that notifications are dispatched to one of a number of possible consumers that are part of the same logical subscriber or parallelized application.
```

This can be done by calling token create with the same parameters as originally. 
If the parameters used are not available they can be extracted from the token.
As the token string is a JWT (JSON Web Token), it can be decoded to extract the original information used to create the token by splitting it into 3 parts (on ".") and doing a base64 decode on the first substring.
This way, information like the subscription name can be extracted and the create token REST point can be called again, all on the client side.
The {{< product-c8y-iot >}} microservice Java SDK [TokenApi](https://github.com/SoftwareAG/cumulocity-clients-java/blob/develop/java-client/src/main/java/com/cumulocity/sdk/client/messaging/notifications/TokenApi.java) class contains a public refresh method which is implemented purely on the client side.

### Shared subscription

There can be multiple subscription on a managed object, each receiving filtered notifications as specified by their individual subscriptions.
In order to scale, shared subscriptions are required so that notifications are dispatched to one of a number of possible consumers that are part of the same logical subscriber or parallelised application.

> This can be achieved by creating a token with a subscription and subscriber name for the scalable application but with an optional boolean "isShared" request parameter set to true in the token create request.

Reviewer: ```suggestion
This can be achieved by creating a token with a subscription and subscriber name for the scalable application with the optional boolean request parameter `isShared` set to true in the token create request.
```

Notifications (messages) will be distributed among all connections to the Notification 2.0 web socket endpoint.

> These all use a token for the same subscription and the subscriber and both the same subscription name, either by using the same shared token or using the same subscription name and subscriber when generating per instance tokens.

Reviewer: "These" -> "They"
"same subscription and the subscriber and both the same subscription name" - what is meant?
are "per instance tokens" special tokens?

> The subscription name defines the "topic" messages are published on, while the subscriber identifies the backend (north side) application that's usually used.

Reviewer: why is "topic" in quotes?

Author: removed quotes here

The application can consist of more than one instance running in parallel in the shared use case.

Notifications will be delivered in order with respect to the notification generating device.
The notifications will be delivered to the same application instance, except when a new instance or a failure changes the application topology.
Then it is necessary to re-distribute the devices to currently running instances.
In order to aid this assignment, the subscriber instance can specify a "consumer name" when connecting to the web socket endpoint.
The same token, or one that was generated in a similar way from subscription name and subscriber, is used as the token query string argument.

> However, a different consumer name is passed as the "consumer" query string argument.

Reviewer: ```suggestion
However, a different consumer name is passed as the `consumer` query string argument.
```

For example, instance 1 of the subscriber microservice could pass in `notification2/consumer?token=xyz&consumer=instance1` while instance 2 could use `notification2/consumer?token=XYZ&consumer=instance2`.
Currently, determining the instance ID of a microservice replica is not supported by the {{< product-c8y-iot >}} microservice API. An external application with named instances can be used instead.

### Volatile subscriptions

> When subscribing, it is possible to pass in an optional boolean "isVolatile" query parameter set to true.

Reviewer: ```suggestion
When subscribing, it is possible to pass in an optional boolean `isVolatile` query parameter set to true.
```

> Note that there is no need to mark a subscription as shared - only the token is marked shared. 

Reviewer: ```suggestion
Note that there is no need to mark a subscription as shared - only the token is marked as shared. 
```

> Both token and subscriptions have isVolatile but only token has both isVolatile and isShared.

Reviewer: ```suggestion
Both the token and the subscriptions have `isVolatile` but only the token has both `isVolatile` and `isShared`.
```

This changes the subscription to not persist notifications for the named subscription.
They are effectively only buffered in memory and can be discarded if they are not consumed quickly enough or on node failure.
Note that there can be both volatile and ordinary (non-volatile or persistent) subscriptions on a managed object with the same subscription name.
These count as separate subscriptions and can be consumed by a subscriber using a token with the corresponding `isVolatile` equal to true or false.   

### Unsubscribing a subscriber

Once a subscription is made, notifications will be kept until consumed by all subscribers who have previously connected to the subscription.
For non-volatile subscriptions, this can result in notifications remaining in storage if never consumed by the application.

> These will be deleted if a tenant is deleted but otherwise can take up considerable space in permanent storage for high frequency notification sources.

Reviewer: ```suggestion
They will be deleted if a tenant is deleted but otherwise can take up considerable space in permanent storage for high frequency notification sources.
```

It is therefore advisable to unsubscribe a subscriber that will never run again.
A separate REST endpoint is available for this: <kbd>/notification2/unsubscibe</kbd>.

> It has a non-optional query parameter "token".

Reviewer: ```suggestion
It has a non-optional query parameter `token`.
```

The token is the same as you would use to connect to the web socket endpoint to consume notifications.

> Note that there is no explicit subscribe subcriber using a token.

Reviewer: "subscribe subscriber" - is this correct?

> Instead this happens when you first connect a Web Socket with a token for the subscription name and subscriber.

Reviewer: ```suggestion
Instead this happens when you first connect a web socket with a token for the subscription name and subscriber.
```

> To unsubscribe an application, however, is an explicit act using a similar or the original token token.

Reviewer: ```suggestion
However, unsubscribing an application is an explicit act using the original or a similar token.
```

Unsubscribing should be infrequent, for example when deleting an application or during development when testing completes, 
as typically one wants messages to persist even when no consumer is running.

> Only if no consumer will ever run again should an unsubscribe subscriber be necessary.

Reviewer: ```suggestion
Only if no consumer will ever run again should unsubscribing a subscriber be necessary.
```

It is also possible to unsubscribe a subscriber on an open consumer web socket connection.
To do so, send "unsubscribe-subscriber" instead of a message acknowledgement identifier from your web socket client to the service.
The service will then unusbscribe the subscriber and close the connection.

> It's not possible to check if the unsubscribe succeeded as the connection always closes so this way of unsubscribing is mostly for testing.

Reviewer: ```suggestion
It's not possible to check if the unsubscribe operation succeeded as the connection always closes so this way of unsubscribing is mostly for testing.
```

For Java developers, the API and the protocol have been wrapped up as an open Java API and a sample WebSocket client application.

There is a sample microservice available in the [cumulocity-examples repository](https://github.com/SoftwareAG/cumulocity-examples/tree/develop/hello-world-notification-microservice), so Java developers do not need to code to the following protocol specification directly.

### Token expiration

> When creating a token, an expiration time must be given in minutes of validity from when the token was created.

Reviewer 3: I think we can skip this part: "When creating a token, an expiration time must be given in minutes of validity from when the token was created.
This security feature limits the potential damage due to leaking of a token." 
and just talk about refresh only.

Author: I think it's a reminder of why we are even talking about it. No harm?

Reviewer 2: Yeah, I don't see any harm in leaving this in.
